File: app/src/main/java/com/josycom/mayorjay/flowoverstack/viewmodel/CustomSearchViewModelFactory.java
package com.josycom.mayorjay.flowoverstack.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.josycom.mayorjay.flowoverstack.repository.SearchRepository;

import javax.inject.Inject;

public class CustomSearchViewModelFactory implements ViewModelProvider.Factory {

    private SearchRepository searchRepository;

   @Inject
    public CustomSearchViewModelFactory(SearchRepository searchRepository) {
        this.searchRepository = searchRepository;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new SearchViewModel(searchRepository);
    }
}

File: app/src/main/java/com/josycom/mayorjay/flowoverstack/viewmodel/CustomSearchViewModelFactory.kt
package com.josycom.mayorjay.flowoverstack.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.josycom.mayorjay.flowoverstack.data.repository.SearchRepository
import javax.inject.Inject

class CustomSearchViewModelFactory @Inject constructor(private val searchRepository: SearchRepository) : ViewModelProvider.Factory {

    private var page = 0
    private var pageSize = 0

    fun setInputs(page: Int, pageSize: Int) {
        this.page = page
        this.pageSize = pageSize
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        when {
            modelClass.isAssignableFrom(SearchViewModel::class.java) -> return SearchViewModel(
                page,
                pageSize,
                searchRepository
            ) as T
            else -> throw IllegalArgumentException("Wrong Class Provided $modelClass")
        }
    }
}